### PR TITLE
Moz lang headers

### PR DESCRIPTION
**Pull request by phlax**
Adds support for moz_lang headers, and tag comments

### CODE REVIEW

File: translate/storage/test_mozilla_lang.py [hunk 1]
```diff
@@ -121,3 +127,59 @@ def test_ok_translations(self, ok, target, istranslated):
...
+                "## NOTE: foo\n"
+                "## TAG: fooled_you ##\n"
+                ";Source\n"
+                "Target\n"
+                "\n\n")
+        store = self.StoreClass.parsestring(lang)
+        assert "## TAG: fooled_you ##" not in store.getlangheaders()
+
+    def test_tag_comments(self):
+        """Ensure we can handle and preserve miltiline comments"""
```
Comment (phlax) [inline translate/storage/test_mozilla_lang.py]: need to change these docstrings

File: translate/storage/test_mozilla_lang.py [hunk 2]
```diff
@@ -121,3 +127,59 @@ def test_ok_translations(self, ok, target, istranslated):
...
+        lang = ("## active ##\n"
+                "# First comment\n"
+                "## TAG: important_tag\n"
+                "# Second comment\n"
+                "# Third comment\n"
+                "## TAG: another_important_tag\n"
+                ";Source\n"
+                "Target\n"
+                "\n\n")
```
Comment (unho) [inline translate/storage/test_mozilla_lang.py]: When it comes to snippets of any format I find preferable to use triple quoted strings so we can copy-paste verbatim snippets.

File: translate/storage/mozilla_lang.py [hunk 3]
```diff
@@ -116,6 +130,13 @@ def parse(self, lines):
...
             out.write(b"## active ##\n")
+        for header in self._headers:
+            out.write(bytes("%s\n" % header))
+        if self._headers:
+            out.write(b"\n\n")
         for unit in self.units:
             out.write(six.text_type(unit).encode('utf-8'))
             out.write(b"\n\n\n")
+
+    def getlangheaders(self):
```
Comment (unho) [inline translate/storage/mozilla_lang.py]: I wonder if instead of `self._headers` it is better to just have `self.headers`. That way we can skip this extra method.

File: translate/storage/mozilla_lang.py [hunk 4]
```diff
@@ -116,6 +130,13 @@ def parse(self, lines):
     def serialize(self, out):
         if self.is_active or self.mark_active:
             out.write(b"## active ##\n")
+        for header in self._headers:
```
Comment (unho) [inline translate/storage/mozilla_lang.py]: I would use `line` and `self.header`. It is not like you have multiple headers at the beginning of the file, just 1 header with multiple lines.

File: translate/storage/mozilla_lang.py (hunk 3 above)
Comment (phlax) [inline translate/storage/mozilla_lang.py]: i want to explicitly make this the public method - afaict the only other format that has headers implemented is po - and in that case its called getpoheaders (iirc) - because they are format-specific

Comment (phlax) [inline translate/storage/mozilla_lang.py]: we might want to think about a generic approach to handling headers - but unless we do i think it should be explicit

File: translate/storage/mozilla_lang.py (hunk 4 above)
Comment (phlax) [inline translate/storage/mozilla_lang.py]: im not really following your point here

File: translate/storage/test_mozilla_lang.py [hunk 5]
```diff
@@ -121,3 +127,59 @@ def test_ok_translations(self, ok, target, istranslated):
...
+        lang = ("## active ##\n"
+                "## some_tag ##\n"
+                "## another_tag ##\n"
+                "## NOTE: foo\n"
+                "## TAG: fooled_you ##\n"
+                ";Source\n"
+                "Target\n"
+                "\n\n")
+        store = self.StoreClass.parsestring(lang)
+        assert "## TAG: fooled_you ##" not in store.getlangheaders()
```
Comment (unho) [inline translate/storage/test_mozilla_lang.py]: It would be nice to test too `MAX_LENGTH`

Comment (unho) [inline translate/storage/test_mozilla_lang.py]: I guess it doesn't belong to this PR.

File: translate/storage/test_mozilla_lang.py [hunk 6]
```diff
@@ -121,3 +127,59 @@ def test_ok_translations(self, ok, target, istranslated):
...
+            == ("## active ##\n"
+                "## some_tag ##\n"
+                "## another_tag ##\n"
+                "## NOTE: foo\n"
+                "\n\n"
+                ";Source\n"
+                "Target\n"
+                "\n\n"))
+
+    def test_not_headers(self):
```
Comment (unho) [inline translate/storage/test_mozilla_lang.py]: I would add another test to ensure that

```
lang = """## active ##
## some_tag ##
## another_tag ##
## NOTE: foo
## TAG: fooled_you ##
# Note for string
;Source
Target


"""
```

correctly parses the unit comment.

File: translate/storage/mozilla_lang.py (hunk 3 above)
Comment (unho) [inline translate/storage/mozilla_lang.py]: I am against this specific things. They will just make more difficult to get to a common API. I would say have `headers` always, and if the format doesn't support them just have empty list and output empty string. After all it is what this implementation is doing.

File: translate/storage/mozilla_lang.py (hunk 4 above)
Comment (unho) [inline translate/storage/mozilla_lang.py]: The file has a single header, not multiple headers. That single header can have multiple lines. So I suggest to alter the variables names to reflect that.

Comment (phlax) [inline translate/storage/mozilla_lang.py]: thats not correct i dont believe

File: translate/storage/mozilla_lang.py (hunk 3 above)
Comment (phlax) [inline translate/storage/mozilla_lang.py]: great - please raise a PR to address these more fundamental concerns and we can think about it then

File: translate/storage/test_mozilla_lang.py (hunk 2 above)
Comment (phlax) [inline translate/storage/test_mozilla_lang.py]: personally i really dislike triple quotes in methods etc- but can we focus on the important things please

**Comment by unho** [comment]
Despite I would love the concerns addressed, lgtm given that this is just a patch to keep going.

File: translate/storage/mozilla_lang.py (hunk 4 above)
Comment (dwaynebailey) [inline translate/storage/mozilla_lang.py]: I guess the distinction is headerline (a single line in a header) vs header (a group of items making up a header) vs headers (a number of header blocks).  Tbh I'm not that fussed

File: translate/storage/test_mozilla_lang.py [hunk 7]
```diff
@@ -38,6 +40,10 @@ def test_comments(self):
         unit.addnote("Hello")
         assert str(unit).find("Hello") == 2
         assert str(unit).find("# Hello") == 0
+        unit.addnote("# TAG: goodbye")
```
Comment (dwaynebailey) [inline translate/storage/test_mozilla_lang.py]: Ideally I'd test this on its own or adapt the comment

**Comment by dwaynebailey** [comment]
Some nits, otherwise lgtm

File: translate/storage/test_mozilla_lang.py (hunk 6 above)
Comment (phlax) [inline translate/storage/test_mozilla_lang.py]: this is addressed

File: translate/storage/test_mozilla_lang.py (hunk 2 above)
Comment (unho) [inline translate/storage/test_mozilla_lang.py]: I dislike them too, but in this specific scenario the advantages are greater than the drawbacks and my personal opinion against them.

Specifically using triple quotes you don't have to escape everything, so you can copy-paste easily and you don't lose plenty of time escaping. Also the snippet looks like like in real life so it is considerably easier for humans to understand it, and therefore understand what the test is about. Just my opinion.

File: translate/storage/test_mozilla_lang.py (hunk 7 above)
Comment (phlax) [inline translate/storage/test_mozilla_lang.py]: i put together because the behaviour changed - and i felt like it should go here - ive updated the comment

File: translate/storage/test_mozilla_lang.py [hunk 8]
```diff
@@ -121,3 +127,87 @@ def test_ok_translations(self, ok, target, istranslated):
...
+                "## TAG: another_important_tag\n"
+                "# Another comment\n"
+                ";Source\n"
+                "Target\n"
+                "\n\n")
+        store = self.StoreClass.parsestring(lang)
+        assert not store.getlangheaders()
+        assert (
+            "First comment"
+            in store.units[0].getnotes(origin="developer").split("\n"))
```
Comment (unho) [inline translate/storage/test_mozilla_lang.py]: Is it not better to get all the dev notes for the first units and store in variable? Considerable shorter asserts I would say.

Comment (phlax) [inline translate/storage/test_mozilla_lang.py]: i think this is more explicit
